### PR TITLE
# feat: Create GET /api/challenges endpoint with CORS support (IC-02, Issue #53)

### PR DESCRIPTION
**Pull request by sserurich**
# feat: Create GET /api/challenges endpoint with CORS support (IC-02, Issue #53 #79 )

## Description

Implement a public-facing API endpoint to fetch active challenges for integration with the LearnWorlds portal. The endpoint enables external platforms to display challenge cards to learners without requiring authentication.

---

### New Files

- `app/api/challenges/route.ts` — Public API endpoint for fetching challenges with CORS headers
- `tests/challenges-api.spec.ts` — Comprehensive Playwright test suite (11 tests)

### Modified Files

- `middleware.ts` — Excluded `/api/challenges` from auth middleware (public endpoint)
- `.env.local` — Added environment variables for CORS origin and signup URL configuration

---

#### Query Parameters

- `status` (optional): Filter challenges by status (`setup`, `open`, `active`, `completed`)
- Defaults to `open` if omitted or invalid

---

### CORS Support

- Handles `OPTIONS` preflight requests (204 status)
- Respects `LEARNWORLDS_ALLOWED_ORIGIN` environment variable
- Includes:
  - `Access-Control-Allow-Methods`
  - `Access-Control-Allow-Headers`
  - `Access-Control-Max-Age`
- Blocks CORS if origin does not match

---

### Response Format

```json
{
  "challenges": [
    {
      "id": "uuid",
      "title": "Challenge Name",
      "short_description": "Brief overview...",
      "cover_image_url": null,
      "challenge_type": "global",
      "tags": [],
      "prize_amount": null,
      "deadline": null,
      "teams_registered_count": 12,
      "participant_signup_url": "http://localhost:3000/participant/event/{id}"
    }
  ]
}
```

---

## Error Handling

- Returns `400` for invalid status parameters with descriptive error message
- Returns `500` for database/server errors
- Applies CORS headers to all responses

---

## Database Query

- Left joins `events` to count registered teams per challenge
- Groups by event ID and aggregates team count
- Orders results by creation date

---

#### Install dependencies

```bash
npm install
```

#### Start dev server (if not running)

```bash
npm run dev
```

#### Run all challenge API tests

```bash
npx playwright test tests/challenges-api.spec.ts
```

#### Run with specific browser

```bash
npx playwright test tests/challenges-api.spec.ts --project=chromium
```

---

#### Test basic endpoint

```bash
curl -s http://localhost:3000/api/challenges | jq .
```

#### Test with specific status

```bash
curl -s "http://localhost:3000/api/challenges?status=active" | jq .
```

#### Test invalid status (should return 400)

```bash
curl -s "http://localhost:3000/api/challenges?status=invalid" | jq .
```

#### Test CORS preflight

```bash
curl -i -X OPTIONS http://localhost:3000/api/challenges \
  -H "Origin: https://www.learnworlds.com"
```

#### Test CORS with LearnWorlds origin
```bash
curl -s http://localhost:3000/api/challenges \
  -H "Origin: https://www.learnworlds.com" | jq .
```
#### Challenges Detail API (GET /api/challenges/[id])
```bash
curl -s http://localhost:3000/api/challenges/bda1e2c9-f71d-4846-b29d-652b20bd8ab9 | jq .
```
---

## Environment Configuration

Add to `.env.local`:

```env
LEARNWORLDS_ALLOWED_ORIGIN=https://your-learnworlds-domain.com
PUBLIC_PARTICIPANT_SIGNUP_BASE_URL=http://localhost:3000
```

---
## Related Issues 
- #79 
## 🧪 Test Coverage

✅ **11 Playwright tests — All passing**

- Default status filter (`open`)
- Filter by different statuses (`active`, `completed`)
- Invalid status returns `400` error
- Response shape validation
- Field types validation
- Team count aggregation
- `OPTIONS` preflight CORS headers
- `GET` response includes CORS headers
- CORS origin validation
- Public access (no auth required)
- Query parameter trimming

---

## ✅ Acceptance Criteria

- Endpoint returns only `open` challenges by default
- CORS headers verified for LearnWorlds domain
- Database join includes team count per challenge
- Public endpoint (no JWT required)
- Status filtering works for valid statuses
- Invalid status returns `400` error with message
- Participant signup URLs are properly formatted
- All tests passing in CI/CD

---

## ✅ Checklist

- [ ] Code follows project conventions
- [ ] All tests passing (11/11)
- [ ] No TypeScript errors
- [ ] Environment variables documented
- [ ] CORS properly configured
- [ ] Middleware updated for public access
- [ ] Ready for Pull Request review

### CODE REVIEW

**Comment by sserurich** [comment]
Might need to update the implementation to use challenges #52 (has all the necessary fields returned) after it's merged instead of events

**Comment by bofengb** [comment]
# PR Review: `feature/api-challenges-cors`

Hey! Thanks for putting this together — the TDD approach (tests first, then implementation, then refactor) is really well done, and the CORS handling is solid. I think this endpoint is exactly what we need for the LearnWorlds integration.

After reviewing all three competition/challenges PRs together, I want to share some thoughts on how we can align this with the broader feature direction. I think with a few adjustments, this PR becomes the foundation for our public-facing API.

---

## 1. The Goal

We're building a **competitions feature** where certain events can be "promoted" to competitions with extra metadata (prize, tags, cover image, deadline, country, etc.). The architecture we're converging on is:

- A `competitions` table that has a **1:1 relationship with `events`** (via a unique `event_id` FK). It stores only the extra metadata — it does NOT duplicate `status`, `organization_id`, or `max_team_size` from events, because those are inherited.
- Your `/api/challenges` endpoint becomes the **public-facing API** that joins `events + competitions` to serve data to LearnWorlds.
- All existing scoring, teams, criteria, and judge assignment logic continues to work through `events` — no duplication needed.

This means your endpoint is the right approach — "challenges" are a **read-only view** over events+competitions, not a separate entity.

---

## 2. Issues in the Current PR

### 2.1 Hardcoded Placeholder Fields

The response currently returns hardcoded nulls/defaults for fields that will exist in the `competitions` table:

```typescript
cover_image_url: null,
challenge_type: 'global',
tags: [],
prize_amount: null,
deadline: null,
```

Once the `competitions` table exists, these should come from a JOIN with real data.

### 2.2 Query Doesn't Join with Competitions

The current query only reads from `events`:

```typescript
const challenges = await db
  .select({ ... })
  .from(events)
  .leftJoin(teams, eq(teams.eventId, events.id))
  .where(eq(events.status, status))
  .groupBy(events.id, events.name, events.description)
  .orderBy(events.createdAt);
```

It should also LEFT JOIN with `competitions` to pull in the extra metadata, and ideally only return events that HAVE a competition record (i.e., INNER JOIN or filter for `competitions.id IS NOT NULL`).

### 2.3 No Pagination

The endpoint returns all matching events. As the platform grows, this could return a very large payload. Adding basic `limit`/`offset` support would be good.

### 2.4 Minor: SQL Type Cast

```typescript
teamsRegisteredCount: sql<string>`count(${teams.id})`,
```

The `sql<string>` annotation then `Number()` conversion works but is slightly roundabout. You could use `sql<number>` with `::int` cast.

### 2.5 Critical: Public Exposure of `setup` Status Events

The valid status whitelist includes `setup`:

```typescript
const VALID_STATUSES = ['setup', 'open', 'active', 'completed'];
```

This means `GET /api/challenges?status=setup` returns draft/setup events publicly. This contradicts the RLS policy in `feat/IC-01-challenges-table` which explicitly filters `e.status != 'setup'` from public view, and the detail endpoint's draft filtering (section 3.4 below) which returns 404 for setup events.

Draft events are not ready for public consumption — they should not be queryable via the public API.

---

## 3. How to Fix

### 3.1 Update the Query to JOIN with Competitions

Once the `competitions` table is available in the Drizzle schema (see the other PR for the table definition), update the query.

Also, remove `'setup'` from the valid statuses to prevent exposing draft events publicly (consistent with IC-01's RLS policy and the detail endpoint's filtering):

```typescript
// In app/api/challenges/route.ts

import { events, teams, competitions } from '@/lib/db/schema';
import { eq, sql, and, isNotNull } from 'drizzle-orm';

// Remove 'setup' from valid statuses — draft events should not be public
const VALID_STATUSES = ['open', 'active', 'completed'];

// Inside the GET handler, replace the current query with:
const challenges = await db
  .select({
    id: events.id,
    title: sql<string>`COALESCE(${competitions.title}, ${events.name})`,
    shortDescription: competitions.shortDescription,
    coverImageUrl: competitions.coverImageUrl,
    challengeType: competitions.challengeType,
    tags: competitions.tags,
    prize: competitions.prize,
    deadline: competitions.deadline,
    country: competitions.country,
    participantSignupUrl: competitions.participantSignupUrl,
    teamsRegisteredCount: sql<number>`count(${teams.id})::int`,
  })
  .from(events)
  .innerJoin(competitions, eq(competitions.eventId, events.id))
  .leftJoin(teams, eq(teams.eventId, events.id))
  .where(eq(events.status, status))
  .groupBy(
    events.id,
    events.name,
    competitions.title,
    competitions.shortDescription,
    competitions.coverImageUrl,
    competitions.challengeType,
    competitions.tags,
    competitions.prize,
    competitions.deadline,
    competitions.country,
    competitions.participantSignupUrl
  )
  .orderBy(events.createdAt);
```

Key changes:
- `INNER JOIN competitions` — only returns events that have been promoted to competitions
- `COALESCE(competitions.title, events.name)` — uses competition title if set, falls back to event name
- All placeholder fields now come from real database columns
- `::int` cast on count for cleaner typing

### 3.2 Update the Response Mapping

```typescript
const responsePayload = {
  challenges: challenges.map((challenge) => ({
    id: challenge.id,
    title: challenge.title,
    short_description: challenge.shortDescription || null,
    cover_image_url: challenge.coverImageUrl || null,
    challenge_type: challenge.challengeType || 'global',
    tags: challenge.tags || [],
    prize_amount: challenge.prize || null,
    deadline: challenge.deadline || null,
    teams_registered_count: challenge.teamsRegisteredCount,
    country: challenge.country || null,
    participant_signup_url:
      challenge.participantSignupUrl ||
      `${participantBaseUrl}/participant/event/${challenge.id}`,
  })),
};
```

### 3.3 Add Basic Pagination (Optional but Recommended)

```typescript
// Parse pagination params
const limit = Math.min(parseInt(searchParams.get('limit') || '50', 10), 100);
const offset = parseInt(searchParams.get('offset') || '0', 10);

// Add to query chain (before .orderBy):
// ...existing query...
  .orderBy(events.createdAt)
  .limit(limit)
  .offset(offset);

// Include in response:
const responsePayload = {
  challenges: challenges.map(/* ... */),
  pagination: { limit, offset, count: challenges.length },
};
```

### 3.4 Add the Detail Endpoint (`/api/challenges/[id]`)

This PR is also the right home for the `/api/challenges/[id]` detail endpoint. There's a separate PR (`ic-03-api-challenges-id`) that implements this, but it should be closed and the detail route built here instead — both endpoints are public-facing routes in the same `/api/challenges/` namespace serving the same LearnWorlds integration, so they should share CORS infrastructure, middleware exemptions, and response conventions.

Add `app/api/challenges/[id]/route.ts` with:
- **INNER JOIN with `competitions`** — same query pattern as the list endpoint above, filtered by event ID
- **Shared `getCorsHeaders()`** — reuse the same CORS function (consider extracting it to a shared util if it isn't already)
- **`OPTIONS` handler** — same preflight support as the list endpoint
- **snake_case response format** — matching the list endpoint's LearnWorlds convention
- **UUID validation** — the `ic-03` PR has good UUID validation logic (`/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i`) that can be carried forward
- **Draft filtering** — events in `setup` status should return 404 (the INNER JOIN with `competitions` + RLS policy handles this at the data layer, but an explicit check is good defense-in-depth)

The middleware exclusion you already have (`api/challenges` prefix) covers `/api/challenges/[id]` for free — no additional middleware changes needed.

### 3.5 Update Tests

The test expectations should be updated to verify that real data comes from the competitions table rather than checking for null placeholders. The existing test structure is great — just update the expected values once the JOIN is in place. Also add tests for the new detail endpoint — UUID validation, 404 for non-existent/setup events, CORS headers on detail responses.

---

## 4. Additional Notes

- **CORS implementation is excellent** — the whitelist-based origin check with `Vary: Origin` header is exactly right. No changes needed there.
- **Middleware exclusion is correct** — `/api/challenges` should remain public.
- **Test coverage is impressive** — 11 tests covering happy paths, validation, CORS preflight, and edge cases. This is the standard we should aim for across the project.
- **Environment variables** (`LEARNWORLDS_ALLOWED_ORIGIN`, `PUBLIC_PARTICIPANT_SIGNUP_BASE_URL`) are well-designed with sensible fallbacks.

### A Note on Naming

You'll notice "challenges" and "competitions" are used across the different PRs. They refer to the same concept — an event with extra external-facing metadata. The table is being named `competitions` (in the `feat/IC-01-challenges-table` PR), while your endpoint stays as `/api/challenges` since that's the LearnWorlds-facing API term. Both names are fine — it's just an internal table name vs. external API name.

### PR Dependency & Merge Order

This PR depends on the `feat/IC-01-challenges-table` PR merging first (which creates the `competitions` table and Drizzle schema). Once that's in, you can update your query to JOIN with it. The current PR can be reviewed and approved in parallel, but should merge after the schema PR.

1. `feat/IC-01-challenges-table` (schema + migration) — **ready to merge**
2. This PR (public API endpoint + detail endpoint from `ic-03`)
3. `participant-enhanced` (admin UI, internal API routes, auth guards, participant views)
4. `ic-03-api-challenges-id` — **close this PR** (detail endpoint folded into this PR per section 3.4)

---

This PR is in great shape. The main change is just updating the query to JOIN with the `competitions` table once it's available. Looking forward to seeing this land!

**Comment by sserurich** [comment]
## PR Updates Summary

Implemented all requested updates in this PR. Work was structured in two phases:
1) Enhancements to the **list endpoint**
2) Implementation of the **detail endpoint**
Both phases include corresponding test coverage.

---

### 1) Challenges List API Updates (`GET /api/challenges`)

- Updated query to include **competition metadata via join**, enabling richer challenge responses  
- Added **pagination support** with `limit` and `offset` query parameters  
- Included **pagination metadata** in the response payload  

**Validation Improvements**
- Invalid `status` now returns **400**  
- `setup` status is explicitly **rejected** (contract enforced)  
- Pagination params are **validated and sanitized** (defaults and bounds covered by tests)  

**Additional Enhancements**
- Included additional response fields (e.g., `country`)  
- Maintained **public access + CORS compatibility** for LearnWorlds integration  

---

### 2) Challenges Detail API (`GET /api/challenges/[id]`)

- Added new **public detail endpoint** for fetching a single challenge by ID  

**Validation & Error Handling**
- UUID format validation → returns **400** for invalid IDs  
- Returns **404** for:
  - Non-existent challenges  
  - `setup` challenges (visibility rules enforced)  

**Other Enhancements**
- Added **OPTIONS preflight support** with consistent CORS behavior  
- Response structure follows contract under `challenge` with **snake_case fields**  

---

### 3) Shared API Utility Refactor

- Extracted reusable helpers into `utils.ts`:
  - CORS header generation  
  - UUID validation  

- Refactored both list and detail routes to use shared utilities:
  - Ensures **consistent behavior**
  - Improves **maintainability**

---

### 4) Test Coverage Added / Updated

#### List Endpoint Tests (`challenges-api.spec.ts`)
- Pagination:
  - Defaults  
  - Bounds / clamping  
  - Invalid input handling  
- Status validation:
  - `setup` rejection enforced  
- Response contract checks:
  - Newly added fields (including `country`)  
- Existing:
  - Public access  
  - CORS behavior  

#### Detail Endpoint Tests (`challenges-api.spec.ts`) 
- **200** → Valid ID with correct payload shape  
- Ensures **snake_case fields** are present  
- **400** → Invalid UUID format  
- **404** → Non-existent ID  
- **204** → OPTIONS preflight with CORS headers  

---

**Review by AngelVargas1009 — COMMENTED** [review]
I went through the updates and this looks a lot more aligned with the intended design now. The join with competitions, removal of setup status, and added pagination all make a big difference. The response structure and shared utilities also look clean and consistent.

I also think the test coverage is strong. It is nice to see validation, pagination, and CORS behavior all being tested. That definitely adds confidence to the endpoint.

One thing to note is that the Playwright checks are currently failing, so I would want to make sure those are passing before this gets merged. Since a lot of behavior was updated, it might just be test expectations needing to be adjusted, but it is worth checking it out. 

Other than that, everything looks much closer to the overall feature direction now, especially with the competitions based approach and the detail endpoint being included here.

Overall this looks like solid progress. I would just fix the failing checks and then it should be in a good spot to merge.

**Review by zsin520 — COMMENTED** [review]
I have several thoughts which relate to FR-14 (Standardized Error Handling) which is currently waiting to be merged, PR #103.

Currently, the error responses in these routes are returning { error: "..." }. FR-14 was designed to address this, we would need to swap these out for the new sendApiError() utility, which enforces the { error_code, error_message } JSON structure. There are several areas for changes in `app/api/challenges/[id]/route.ts` and `app/api/challenges/route.ts`.

`sendApiError` does not currently accept custom headers. We may need to upgrade the utility to accept headers. 

FR-14 has changed the structure of the error message, so the testing for this PR would have to be changed slightly. The following lines would have to be changed:
```
// original
expect(data).toHaveProperty('error');
expect(data.error).toContain('Invalid status parameter');

// new
expect(data).toHaveProperty('error_code');
expect(data.error_message).toContain('Invalid status parameter');
```

File: app/api/admin/metrics/route.ts [hunk 1]
```diff
@@ -15,23 +15,20 @@ import { eq } from 'drizzle-orm';
...
     }
 
     // Query total learners (participants)
-    const learners = await db
-      .select()
-      .from(users)
-      .where(eq(users.role, 'participant'));
-    
-// TODO: Replace placeholder metrics with real analytics queries from the database
+    const learners = await db.select().from(users).where(eq(users.role, 'participant'));
```
Comment (8bit-Apollo) [inline app/api/admin/metrics/route.ts:25]: Readability was slightly better before

File: app/api/challenges/[id]/route.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import { NextRequest, NextResponse } from 'next/server';
+import { and, eq, sql } from 'drizzle-orm';
+import { db } from '@/lib/db';
+import { competitions, events, teams } from '@/lib/db/schema';
+import { getCorsHeaders, isValidUuid } from '../utils';
+
+type RouteContext = {
+  params: Promise<{ id: string }>;
```
Comment (8bit-Apollo) [inline app/api/challenges/[id]/route.ts]: is implementing params as a Promise necessary here? Params should already be synchronous so it may be better for clarity and testing purposes to make this a plain object

File: app/api/challenges/route.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+export async function GET(request: NextRequest) {
+  const corsHeaders = getCorsHeaders(request);
+
+  try {
+    const searchParams = request.nextUrl.searchParams;
+    const statusParam = searchParams.get('status')?.trim();
+    const status = resolveStatus(request);
+
+    // Validate that if a status parameter was provided, it's valid
+    if (statusParam && !VALID_STATUSES.includes(statusParam as ChallengeStatus)) {
```
Comment (8bit-Apollo) [inline app/api/challenges/route.ts]: calling valid_statuses here duplicates logic in resolveStatus function. Consider reworking to avoid redundancy.

**Review by 8bit-Apollo — COMMENTED** [review]
The code is well organized, easy to read, and well implemented. Testing is detailed and robust. Some minor fixes could reduce redundancy and improve readability, but these improvements are relatively insignificant. Overall, this is a solid contribution. Good job!

**Review by JBishopCode — COMMENTED** [review]
# PR Review: IC-02 `feature/api-challenges-cors`

Hey, great work getting this to where it is — the competitions JOIN, shared CORS utility, detail endpoint, and pagination all land well and align with the intended architecture. A few things to address before this is ready to merge.

---

## 1. Semi - Blockers

### 1.1 Error Format Doesn't Match IC-20

All error responses in both route files still use `{ error: "..." }`:

```typescript
return NextResponse.json(
  { error: 'Invalid status parameter. Valid values: open, active, completed' },
  { status: 400, headers: corsHeaders }
);
```

PR #103 (IC-20) introduces `sendApiError()` which enforces `{ error_code, error_message }`. Since this PR is merging first, the error format here sets the contract — IC-20 will be responsible for updating these routes to match the new structure when it lands. No changes needed here on your end for this.

Also worth noting for whoever picks up IC-20: `sendApiError()` doesn't currently accept custom headers, so it will need to be extended to support CORS header passthrough before it can be used in these routes.

---

## 2. Issues to Address

### 2.1 `params` Typed as a Promise in the Detail Route

```typescript
type RouteContext = {
  params: Promise<{ id: string }>;
};
```

This was flagged inline. `context.params` in Next.js App Router is a plain synchronous object — typing it as a Promise and `await`-ing it works incidentally but adds confusion and unnecessary async overhead. Change it to:

```typescript
type RouteContext = {
  params: { id: string };
};

// and access directly:
const { id } = context.params;
```

### 2.2 `draft` Status Check Has No Schema Backing

In `/api/challenges/[id]/route.ts`:

```typescript
if (!challenge || challenge.status === 'setup' || challenge.status === 'draft') {
```

`draft` is not a value in `eventStatusEnum` — the only valid statuses are `setup`, `open`, `active`, and `completed`. This check is dead code and should be removed to keep the guard consistent with the actual schema:

```typescript
if (!challenge || challenge.status === 'setup') {
```

### 2.3 Prettier Formatting

There are some inconsistent formatting issues across the new files that don't align with the project's Prettier config. Run the following before merging to fix your Lint and Format Errors:

```bash
npx prettier --write ./
```

### 2.4 `package-lock.json` and `package.json` Changes

There are changes to `package-lock.json` and/or `package.json` that fall outside the scope of this PR. Please revert these to avoid introducing unintended dependency changes:

```bash
git checkout main -- package-lock.json
```
```bash 
git checkout main -- package.json
```

---

## 3. Minor Notes

### 3.1 Unrelated Formatting Change in `metrics/route.ts`

The multi-line query in `app/api/admin/metrics/route.ts` was compacted to a single line, and a previous reviewer already flagged that readability was better before. This change is also out of scope for this PR — revert it.

---

## 4. Local Testing

Pulled this branch, applied the competitions migration locally, and ran the full test suite:

```
63 passed (1.1m)
```

All 63 tests passing across Chromium, Firefox, and WebKit. The CI failures are a database environment issue — the competitions migration isn't being applied against the CI database before Playwright runs. This is a CI config fix, not a code fix. Whoever manages the pipeline should add a migration step before the Playwright job.

---

## 5. PR Dependency and Merge Order Reminder

1. **#77 (IC-01)** — `challenges` table · *already merged* — foundation for all IC work
2. **#93 (IC-05)** — auto-generate `participant_signup_url` trigger · *branches off IC-01; should land before the admin form ships*
3. **This PR #78 (IC-02)** — `GET /api/challenges` public endpoint · *needs IC-01 table*
4. **#116 (IC-03)** — `GET /api/challenges/[id]` detail endpoint · *extends IC-02 API layer*
5. **#104 (IC-04)** — admin challenge management UI · *needs IC-01 schema + IC-02/IC-03 routes*
6. **#100 (IC-06)** — LearnWorlds page + embed container setup · *no code dep, but needs IC-02 to embed against*
7. **#85 (IC-07)** — challenge gallery widget (HTML/CSS/JS) · *consumes `GET /api/challenges`*
8. **IC-08** *(no PR yet)* — client-side search & filter on widget · *extends IC-07*
9. **#84 (IC-09)** — Liquid variable injection into Apply button · *extends IC-07*
10. **#105 (IC-10)** — E2E user flow testing · *needs IC-09 complete*
11. **#97 (IC-19)** — LearnWorlds ingestion foundation schema (`learnworlds_sync_runs`, `learnworlds_raw_payloads`, `learner_progress`) · *prerequisite for IC-11 and IC-13*
12. **#98 (IC-11)** — LearnWorlds ingestion client (OAuth2 auth + fetch) · *branches off IC-19*
13. **IC-12** *(no PR yet)* — tests for IC-11 (auth failure + happy path) · *branches off IC-11*
14. **#99 (IC-13)** — validation, transformation & persistence pipeline · *branches off IC-11 + IC-19*
15. **IC-15** *(no PR yet)* — recommendation engine (rule-based scoring logic) · *needs IC-13's `learner_progress` data*
16. **IC-14** *(no PR yet)* — `GET /api/recommendations/[learner_id]` endpoint · *wraps IC-15 engine*
17. **IC-16** *(no PR yet)* — Playwright + Istanbul JS code coverage · *cross-cutting NFR; best landed after IC-12/IC-13 tests exist*
18. **IC-17** *(no PR yet)* — `GET /admin/metrics` analytics endpoint · *needs recommendation data flowing (IC-14/IC-15)*
19. **#112 (IC-18)** — BGE-M3 AI screening for submission relevance · *standalone AI module; best after ingestion pipeline is stable (IC-13)*
20. **#103 (IC-20)** — centralized API error utility + route refactor (FR-14) · *must land after IC-14 and MD-06 both exist, as it refactors both routes*

Note: IC-05 and IC-02 can be swapped as they don't depend on each other.

---

Fix the items in section 2 and this is ready to merge. The core architecture is solid.

**Comment by sserurich** [comment]
> One thing to note is that the Playwright checks are currently failing, so I would want to make sure those are passing before this gets merged. Since a lot of behavior was updated, it might just be test expectations needing to be adjusted, but it is worth checking it out.

Playwright checks are passing, probably when you ran earlier, you didn't have competition schema locally. It's now merged into this PR from main(back then it was still under another PR which has been merged into main now).  
See attachment and also you can look at the CI
Screenshot 2026-04-09 at 2 17 29 PM

File: app/api/challenges/[id]/route.ts (hunk 2 above)
Comment (sserurich) [reply app/api/challenges/[id]/route.ts]: **Updates applied as described below, thanks for the suggestion!**

I’ve removed the separate `RouteContext` type alias and inlined the type directly in the handler signature to align with the rest of the codebase conventions(can checkout other dynamic routes e.g in **events/[eventid]/ route.ts, teams[teamid]/route.ts** )

I intentionally kept `params` as a `Promise` because this route is running on **Next.js 15.4**, where dynamic route params in Route Handlers are asynchronous. Using:

```ts
{ params }: { params: Promise<{ id: string }> }
```
and then
```ts
const { id } = await params;
```
follows the expected pattern for this version and avoids potential runtime issues from treating `params` as a synchronous object.

For reference:
- Route Handlers: https://nextjs.org/docs/app/api-reference/file-conventions/route  
- Dynamic Routes: https://nextjs.org/docs/app/api-reference/file-conventions/dynamic-routes  

cc @8bit-Apollo

File: app/api/challenges/route.ts (hunk 3 above)
Comment (sserurich) [reply app/api/challenges/route.ts]: Thanks for calling this out. You were right that the status membership check was duplicated, but there was an intentional behavior split before this refactor:

- The resolver normalized invalid or missing values to `open` so the query could still run safely.
- The GET handler separately rejected explicitly invalid user input with a `400`.

So while the validation logic looked redundant, removing the GET check without changing the resolver would have changed behavior by silently treating invalid input as `open`.

To address both concerns, I refactored `route.ts` so `resolveStatus` is now the single source of truth. It now returns both:

- Resolved status  
- `hasInvalidStatusParam` flag  

The GET handler now consumes that result and returns `400` based on the flag, removing the duplicated `VALID_STATUSES.includes` check while preserving the original API behavior:

- Missing `status` → defaults to `open`  
- Valid `status` → uses provided value  
- Invalid `status` → returns `400`

cc @8bit-Apollo

**Comment by sserurich** [comment]
> ## 2. Issues to Address
> ### 2.1 `params` Typed as a Promise in the Detail Route
> ```ts
> type RouteContext = {
>   params: Promise<{ id: string }>;
> };
> ```
> 
> This was flagged inline. `context.params` in Next.js App Router is a plain synchronous object — typing it as a Promise and `await`-ing it works incidentally but adds confusion and unnecessary async overhead. Change it to:
> 
> ```ts
> type RouteContext = {
>   params: { id: string };
> };
> 
> // and access directly:
> const { id } = context.params;
> ```
> 
> ### 2.2 `draft` Status Check Has No Schema Backing
> In `/api/challenges/[id]/route.ts`:
> 
> ```ts
> if (!challenge || challenge.status === 'setup' || challenge.status === 'draft') {
> ```
> 
> `draft` is not a value in `eventStatusEnum` — the only valid statuses are `setup`, `open`, `active`, and `completed`. This check is dead code and should be removed to keep the guard consistent with the actual schema:
> 
> ```ts
> if (!challenge || challenge.status === 'setup') {
> ```
> 
> ### 2.3 Prettier Formatting
> There are some inconsistent formatting issues across the new files that don't align with the project's Prettier config. Run the following before merging to fix your Lint and Format Errors:
> 
> ```shell
> npx prettier --write ./
> ```
> 
> ### 2.4 `package-lock.json` and `package.json` Changes
> There are changes to `package-lock.json` and/or `package.json` that fall outside the scope of this PR. Please revert these to avoid introducing unintended dependency changes:
> 
> ```shell
> git checkout main -- package-lock.json
> ```
> 
> ```shell
> git checkout main -- package.json
> ```
> 
> ## 3. Minor Notes
> ### 3.1 Unrelated Formatting Change in `metrics/route.ts`
> The multi-line query in `app/api/admin/metrics/route.ts` was compacted to a single line, and a previous reviewer already flagged that readability was better before. This change is also out of scope for this PR — revert it.

All the issues in this section, have been addressed as shown in the conversation. 2.1 addressed, 2.2 @AngelVargas1009 sent in a commit addressing it, 2.3 has been addressed and 2.4 (package-lock.json` and `package.json) have been overwritten with the versions from the recent main branch. Additionally 3.1 Unrelated Formatting Change in `metrics/route.ts has been replaced with the version from main. cc @JBishopCode

**Comment by JBishopCode** [comment]
> All the issues in this section, have been addressed as shown in the conversation. 2.1 addressed, 2.2 @AngelVargas1009 sent in a commit addressing it, 2.3 has been addressed and 2.4 (package-lock.json`and`package.json) have been overwritten with the versions from the recent main branch. Additionally 3.1 Unrelated Formatting Change in `metrics/route.ts has been replaced with the version from main. cc @JBishopCode

Awesome. As soon as I see the changes come through and get committed, I will merge! @sserurich

File: app/api/admin/metrics/route.ts (hunk 1 above)
Comment (sserurich) [reply app/api/admin/metrics/route.ts:25]: I checked out the current version from main, it's like this. We can go with this the way it is for now(it's something that came in from merging, didn't change it)

**Comment by sserurich** [comment]
> All 63 tests passing across Chromium, Firefox, and WebKit. The CI failures are a database environment issue — the competitions migration isn't being applied against the CI database before Playwright runs. This is a CI config fix, not a code fix. Whoever manages the pipeline should add a migration step before the Playwright job.

For the failing CI tests, I have created a separate issue to update CI to take care of this[[https://github.com/Freddie-Pike/judge-portal/issues/121](https://github.com/Freddie-Pike/judge-portal/issues/121])

**Comment by JBishopCode** [comment]
With new CI update, all Playwright tests are passing! 

PR LGTM!
